Tools/autotest/common.py

```diff
@@ -1467,117 +1467,131 @@ def wait_seconds(self, seconds_to_wait):
+            if msg:
+                return msg.groundspeed
+
+        def validator(value2, target2):
+            if speed_min <= value2 <= speed_max:
                 return True
-        raise WaitGroundSpeedTimeout("Failed to attain groundspeed range")
+            else:
+                return False
+
+        self.wait_and_maintain("Groundspeed", speed_min, lambda: get_groundspeed(timeout), accuracy=(speed_max - speed_min), validator=lambda value2, target2: validator(value2, target2), maintain_target_time=maintain_target_time, timeout=timeout)
 
     def wait_roll(self, roll, accuracy, timeout=30):
         """Wait for a given roll in degrees."""
-        tstart = self.get_sim_time()
-        self.progress("Waiting for roll of %d at %s" % (roll, time.ctime()))
-        while self.get_sim_time_cached() < tstart + timeout:
-            m = self.mav.recv_match(type='ATTITUDE', blocking=True)
-            p = math.degrees(m.pitch)
-            r = math.degrees(m.roll)
-            self.progress("Roll %d Pitch %d" % (r, p))
-            if math.fabs(r - roll) <= accuracy:
-                self.progress("Attained roll %d" % roll)
-                return True
-        raise WaitRollTimeout("Failed to attain roll %d" % roll)
+        def get_roll(timeout2):
+            msg = self.mav.recv_match(type='ATTITUDE', blocking=True, timeout=timeout2)
+            if msg:
+                p = math.degrees(msg.pitch)
+                r = math.degrees(msg.roll)
+                self.progress("Roll %d Pitch %d" % (r, p))
+                return r
+
+        def validator(value2, target2):
+            return math.fabs((value2 - target2 + 180) % 360 - 180) <= accuracy
+
+        self.wait_and_maintain("Roll", roll, lambda: get_roll(timeout), validator=lambda value2, target2: validator(value2, target2), accuracy=accuracy, timeout=timeout)
 
     def wait_pitch(self, pitch, accuracy, timeout=30):
         """Wait for a given pitch in degrees."""
+        def get_pitch(timeout2):
+            msg = self.mav.recv_match(type='ATTITUDE', blocking=True, timeout=timeout2)
+            if msg:
+                p = math.degrees(msg.pitch)
+                r = math.degrees(msg.roll)
+                self.progress("Pitch %d Roll %d" % (p, r))
+                return p
+
+        def validator(value2, target2):
+            return math.fabs((value2 - target2 + 180) % 360 - 180) <= accuracy
+
+        self.wait_and_maintain("Pitch", pitch, lambda: get_pitch(timeout), validator=lambda value2, target2: validator(value2, target2), accuracy=accuracy, timeout=timeout)
+
+    def wait_and_maintain(self, value_name, target, target_retreiver, validator=None, accuracy=2.0, maintain_target_time=0, timeout=30, called_function=None):
         tstart = self.get_sim_time()
-        self.progress("Waiting for pitch of %.02f at %s" % (pitch, time.ctime()))
+        duration_start = None
+        mean_value = 0.0
```

Comment:
This is not the mean value.  This is the sum of the values.  Maybe `sum_of_achieved_values` with the counter renamed to `count_of_achieved_values`